### PR TITLE
Support NULL literal as function argument

### PR DESCRIPTION
**Pull request by dai-chen**
*Issue #, if available:* https://github.com/opendistro-for-elasticsearch/sql/issues/927, https://github.com/opendistro-for-elasticsearch/sql/issues/715

*Description of changes:* 

1. Add undefined type for NULL literal
2. Change JDBC driver to map undefined type to JDBC NULL type

*Documentation:* https://github.com/dai-chen/sql/blob/support-null-literal-as-function-argument/docs/user/general/datatypes.rst#undefined-data-type

*Design:* A new undefined data type is added as narrowest and root of data type tree.

1. `UNDEFINED` represents only the type of NULL literal for now (probably represent some other special literals in future)
2. `UNKNOWN` represents type not supported yet. For example, some new Elasticsearch type is given as function argument which yet to be added to core engine.

With this new type, we're able to support NULL literal. For example, given `ABS(NULL)`, the distance between NULL and INTEGER and DOUBLE are 1 and 2 respectively, so function signature ABS(INTEGER) is used to evaluate its value. Previously NULL literal's type is unknown which caused exception thrown.

null-literal

*Testing:* Add UT and IT. Because SQLite and H2 has different result type for expression with NULL literal, a `NullLiteralIT` is added manually.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/user/general/values.rst [hunk 1]
```diff
@@ -46,6 +46,21 @@ Here is an example::
...
+
+NULL Literal Handling
+---------------------
+
+Because the type of a null literal is unknown, a special data type ``UNDEFINED`` is reserved to allow it be accepted as a valid function argument::
+
+    od> SELECT NULL, NULL = NULL, 1 + NULL, LENGTH(NULL);
+    fetched rows / total rows = 1/1
+    +--------+---------------+------------+----------------+
+    | NULL   | NULL = NULL   | 1 + NULL   | LENGTH(NULL)   |
```
Comment (harold-wang) [inline docs/user/general/values.rst:58]: 1 + NULL should 1 or NULL?
LENGTH(NULL) should be NULL?

Comment (dai-chen) [reply docs/user/general/values.rst:58]: yes, it should be what is documented here. Line 34 explains this. Unless the operator/function is special, expression involving NULL will end with NULL.

**Review by chloe-zh — APPROVED** [review]
LGTM, thanks for the changes!

**Review by harold-wang — APPROVED** [review]
Thanks for the change!

**Pull request merged**